tapsend/send.go

```diff
@@ -364,6 +372,46 @@ func ValidateInputs(inputCommitments tappsbt.InputCommitments,
 	return isFullValueSpend, nil
 }
 
+// ValidateScriptKeysUnique makes sure the outputs of a set of virtual packets
+// don't lead to collisions in and of the trees (e.g. two asset outputs with the
+// same script key in the same anchor output) or with exclusion proofs (e.g. two
+// asset outputs with the same script key in different anchor outputs).
+func ValidateScriptKeysUnique(packets []*tappsbt.VPacket) error {
+	// We'll keep track of all script keys (per top commitment key) that
+	// we've seen so far. Because the script key is the lower tree's
+	// commitment key, it must be unique per top commitment key.
+	scriptKeys := make(map[[32]byte]map[asset.SerializedKey]struct{})
+
+	for _, vPkt := range packets {
+		tapCommitmentKey, err := vPkt.TapCommitmentKey()
+		if err != nil {
+			return fmt.Errorf("unable to determine tap commitment "+
+				"key: %w", err)
+		}
+
+		perAssetMap, ok := scriptKeys[tapCommitmentKey]
+		if !ok {
+			perAssetMap = make(map[asset.SerializedKey]struct{})
+			scriptKeys[tapCommitmentKey] = perAssetMap
+		}
+
+		for idx := range vPkt.Outputs {
+			vOut := vPkt.Outputs[idx]
+
+			// Check if the script key has already been used in this
+			// transaction.
+			scriptKey := asset.ToSerialized(vOut.ScriptKey.PubKey)
+			if _, ok := perAssetMap[scriptKey]; ok {
```

Comment:
Let me see if I understand this:
In theory it would be ok to reuse scriptKeys over different tapCommitmentKeys?

Reply:
Yes, since those would have different paths in the anchored `tapCommitment`.

I think this statement is correct:

"For a given transfer, across all anchor outputs, assets can share a `TapCommitmentKey()`, or an `AssetCommitmentKey()`, but not both."

If both are shared, for any two assets, then we can't generate valid exclusion proofs.